### DISCUSSION TITLE
[Other]  소켓처리(실시간 처리)가 필요한 서비스들 정리

### DISCUSSION THREAD

**Discussion by owel-dev** [42-Tomodachi/ft_transcendence#62] [General]
소켓으로 처리할(실시간 처리가 필요한) 서비스들

> 구현 전에 실시간 처리가 필요한 부분들에 대한 내용을 개략적으로 정리를 해봤습니다.
> 피드백 받고 수정을 거친 후, api명세 작성 및 코드 구현할 예정입니다.
> 수정할 것, 추가할 것, 빼야할 것, 기타 건의사항 등 코멘트 해주시면 반영하도록 하겠습니다.
> 이벤트는 소켓으로 전달되는 이벤트고, 데이터는 백엔드에서 프론트로 넘어가는 데이터입니다.

# 수정사항

- 22일 5시 35분
  - 채팅 목록 페이지 이벤트 통합.  
  - 유저 목록 이벤트 통합.
  - 채팅방 관리자 명령 항목 추가. 
  - 이벤트와 서비스를 구별하여 기재(서비스는 service파일의 함수, 이벤트는 소켓 이벤트).

<br>

# 채팅

---

### 1) 채팅방 생성

한 클라이언트가 채팅방을 만들면 해당 클라이언트는 채팅방 내부로 이동하여야 하며,
모든 클라이언트가 채팅방 목록에서 새로 생긴 채팅방을 실시간으로 확인할 수 있어야 합니다.

### 2) 채팅방 제거

채팅방이 없어지면 모든 클라이언트가 채팅방 목록에서 
채팅방이 없어지는 것을 실시간으로 확인할 수 있어야 합니다.

### 3) 채팅방 상태 변경

채팅방의 상태가 변경되면 모든 클라이언트가 채팅방 목록에서
채팅방의 변경 사항을 실시간으로 확인할 수 있어야 합니다.

- 이벤트: changeRoom

- API

  ```json
  //채팅방 목록
  [
    {
      chatId: number | string,
      roomTitle: string,
      playerCount: number,
      isPublic: boolean,
      isEntered: boolean
    }, 
    // ...
  ]
  ```

### 1) 채팅방 입장, 퇴장

채팅방의 한 클라이언트가 입장하거나 나갈때마다 모든 클라이언트가 채팅방에 접속한 유저를
채팅방 유저 목록에서 실시간으로 확인할 수 있어야 하며,
채팅 목록에서 입장 알림 메시지를 실시간으로 확인할 수 있어야 합니다.

- 이벤트: changeChatUserList

- 서비스: 채팅 등록 Service 호출

- 데이터: 채팅방 유저 목록 데이터

  ```json
  //채팅방에 속한 유저 리스트
  [
    {
      userId: number | string,
      nickname: string,
      role: string(owner | manager| none)
     },
    // ...
  ]
  ```

  

### 2) 채팅 등록

채팅방의 한 클라이언트가 채팅을 등록하면
해당 유저를 차단한 유저를 제외한 모든 클라이언트가 새로 생긴 채팅을 
채팅 목록에서 실시간으로 확인할 수 있어야 합니다.

- 채팅 등록 요청이 오면 서버에서 DB확인하여 각 유저별로 차단목록 확인하여 소켓으로 메세지 정보 전송

- 이벤트: submitChat

- 데이터: 각 유저별로 차단 목록 고려하여 메세지 정보 전송

  ```json
  // 채팅 메세지
  [
    {
      id: number;
      isBroadcast: boolean;
      // ? 는 옵셔널이지만 isBroadcast가 true일때만 from 빼기
      from?: {
        id: number;
        nickname: string;
        profileImage: string;
      };
      message: string;
      fromUser: boolean;
      createdAt: string | number;
    }
  ]
  ```

  

#### 강퇴 명령

채팅방의 관리자가 한 클라이언트를 강퇴하면
채팅방의 모든 클라이언트가 해당 클라이언트가 유저 목록에서 사라지는 것과
강퇴 알림 메세지를 채팅 목록에서 실시간으로 확인할 수 있어야 합니다.
강퇴 당한 클라이언트는 채팅방에서 나가져 채팅 목록 페이지로 이동하여야 하며
’강퇴당했다는 alert 메세지’를 실시간으로 확인할 수 있어야 합니다.

관리자가 아닌 클라이언트가 다른 클라이언트를 강퇴하면
강퇴를 시도한 클라이언트 당사자는
채팅목록에서 “강퇴 실패, 권한 없음” 메세지를 실시간으로 확인할 수 있어야 합니다.

  - 이벤트: 없음.

  - 서비스:

    - 강퇴 성공 시 채팅방 퇴장 service, 채팅 등록 service 호출.
    - 강퇴 실패 시 강퇴 요청한 측 클라이언트에서만 채팅 등록 service 호출

  - 데이터: 성공 시 채팅방 퇴장과 채팅 등록 서비스 호출에 필요한 데이터,
    실패 시 채팅 등록 서비스 호출에 필요한 데이터.

    

#### 관리자 뮤트 명령

채팅방의 관리자가 한 클라이언트를 '관리자 뮤트' 하면
채팅방의 모든 유저의 화면의 유저목록에서 뮤트당한 클라이언트를 식별할 수 있어야 하며
뮤트당한 유저는 채팅 등록 버튼이 비활성화 되어야 한다.

* 이벤트: changeChatUserList
* 데이터: 모든 유저의 정보.

<br>

# 유저 목록

---

### 1) 유저 로그인

특정 클라이언트가 로그인에 성공하면 해당 클라이언트의 상태를 접속중이라 여기며
모든 클라이언트는 해당 유저가 접속중임을 실시간으로 확인할 수 있어야 한다.

### 2) 유저 로그아웃

특정 클라이언트의 토큰이 만료되거나 브라우저가 종료되면 해당 클라이언트의 상태를 종료상태라 여기며
모든 클라이언트는 해당 유저가 종료상태임을 실시간으로 확인할 수 있어야 한다.

### 3) 유저 게임 상태 관리

특정 클라이언트가 게임을 시작하면 해당 클라이언트의 상태를 게임중이라 여기며
모든 클라이언트는 해당 유저가 게임중임을 실시간으로 확인할 수 있어야 한다.

- 이벤트: changeUserStatus
- 데이터: 전체 유저 데이터

<br>

# 게임

---

### 1) 게임방 생성

한 클라이언트가 게임방을 만들면 해당 클라이언트는 게임방 내부로 이동하여야 하며,
모든 클라이언트가 게임방 목록에서 새로 생긴 게임방을 실시간으로 확인할 수 있어야 합니다.

### 2) 게임방 제거

게임방이 없어지면 모든 클라이언트가 게임방 목록에서 
게임방이 없어지는 것을 실시간으로 확인할 수 있어야 합니다.

### 3) 게임방 상태 변경

게임방의 상태가 변경되면 모든 클라이언트가 게임방 목록에서
게임방의 변경 사항을 실시간으로 확인할 수 있어야 합니다.

- 이벤트: changeGameRoom

- 데이터: 전체 게임방 목록

  ```json
  // 게임방 목록 
  [
    {
      gameId: number | string,
      roomTitle: string,
      playerCount: number,
      isPublic: boolean,
      isStart: boolean,
    }, 
    // ...
  ]
  ```

  

### 1) 게임 시작

게임방에 두 번째 클라이언트가 입장하면
게임방의 첫 번째 유저는 해당 유저의 입장을 실시간으로 확인할 수 있어야 하며
즉시 카운트다운이 시작됩니다.

- 카운트 다운 →  게임 화면으로 변환 (프론트에서 처리)
- 카운트 다운이 끝나면 두 클라이언트간의 게임이 시작되어야 하며
  게임 시 시 게임방의 상태가 ‘게임중’으로 변경되어야 합니다. 
  (게임 시작은 소켓으로 처리, 첫 번째 유저가 ‘게임시작' 이벤트를 서버에 보내면 
  서버에서 두 유저에게 ‘게임시작' 이벤트를 브로드캐스트)
    - 이벤트: startGame, changeGameRoom
    - 데이터: 없음.

### 1) 키보드 입력 동기화

각 플레이어의 키보드 위, 아래 입력을 받으면
해당 플레이어의 화면에서 막대 이미지의 위치가 위, 아래로 변경되어야 하며
상대 플레이어의 화면에서도 막대 이미지의 위치 변경이 실시간으로 반영되어야 합니다.

- 이벤트: keyDown, keyUp
- 데이터: 키보드 누른 유저의 데이터

### 2) 플레이어 점수 관리

실점 상황이 되면 득점한 유저의 점수가 올라가야 하며
상대 플레이어의 화면에서도 득점한 유저의 점수가 올라가는 것을 실시간으로 확인할 수 있어야 합니다.

- 매 스코어 게임의 시작과 종료를 소켓으로 처리하여 동기화 해줌으로서 최대한 오차를 낮춤
- 게임 점수 올라가는건 한 스코어 종료 소켓처리 할 때 같이 처리.
  - 이벤트: startScore, endScore
  - 데이터: 이긴 유저의 데이터

### 3) 게임 종료

한 플레이어의 점수가 10점이 되면 게임이 종료되고 모든 플레이어와 관전자들의 화면이 변화해야하며
이긴 플레이어와 진 플레이어의 전적이 기록되어야 한다.

- 이벤트: endGame
- 서비스: 게임방 제거 서비스 호출
- 데이터: 게임 진행한 플레이어 데이터, 게임방 제거 서비스에 필요한 데이터

**Comment by som-syom** [comment]
### 채팅방 리스트
방을 만들고 나가는건 단순 API 요청을 통해 백엔드에 리퀘스트를 보낼 수 있다고 생각하여
실시간 방의 리스트를 업데이트 될때마다 뿌려주면 좋을 듯 합니다.
ex
```
1: 아무나오세요 3명 입장 대기중
2: 비밀방 1명 입장 대기중
3: ㅁㅁㅁ 2명 대기중
```
socket에서 이런 자료를 제일 마지막 최신 상태로 보내주셨었다면
누군가 방을 만들었을 때는 
```
1: 아무나오세요 3명 입장 대기중
2: 비밀방 1명 입장 대기중
3: ㅁㅁㅁ 2명 대기중
4: ㅁㄴㅇㄹㅁㄴㅇㄹ 1명 대기중
```
이렇게 리스트를 보내주시면 됩니다.

추가, 삭제를 보낸다면 현재 이미 만들어져있는 방을 볼때도 애매할 듯 해서 방을 1개씩 관리하는 것이 아닌 리스트로 지속적으로 뿌려주어야 될 것 같습니다!

**Comment by owel-dev** [comment]
@som-syom
_말씀하신 대로 하면
"createRoom", "deleteRoom", "changeRoom" 이렇게 반환하는 소켓의 이벤트를 3개로 나누지 않고
전체 방 목록 데이터를 반환하는 하나의 "Room" 소켓 이벤트로 묶어서 처리할 수 있을 것 같습니다.
수정하여 반영하도록 하겠습니다._

---------

_수정 완료하였습니다. 수정 사항 글 내용에 추가하였습니다._
